### PR TITLE
fix(ivy): align NgModule registration timing with ViewEngine

### PR DESCRIPTION
**Pull request by crisbeto**
Currently in Ivy `NgModule` registration happens when the class is declared, however this is inconsistent with ViewEngine and requires extra generated code. These changes remove the generated code for `registerModuleFactory`, pass the id through to the `ngModuleDef` and do the module registration inside `NgModuleFactory.create`.

This PR resolves FW-1285.

### CODE REVIEW

**Review by mhevery — APPROVED** [review]
`@angular/core` LGTM for the compiler @alxhub should approve.

File: packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts [hunk 1]
```diff
@@ -120,10 +120,16 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
       const bootstrapMeta = this.evaluator.evaluate(expr, forwardRefResolver);
       bootstrapRefs = this.resolveTypeList(expr, bootstrapMeta, name, 'bootstrap');
     }
-
-    const id: Expression|null =
-        ngModule.has('id') ? new WrappedNodeExpr(ngModule.get('id') !) : null;
-
+    let id: string|DynamicValue|null = null;
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts]: I'm actually in favor of not attempting to interpret this value at all, and just passing it through as a `WrappedNodeExpr`, like before. The compiler doesn't need to understand this value.

File: packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts [hunk 2]
```diff
@@ -29,7 +29,7 @@ export interface NgModuleAnalysis {
   metadataStmt: Statement|null;
   declarations: Reference<ClassDeclaration>[];
   exports: Reference<ClassDeclaration>[];
-  id: Expression|null;
+  id: string|DynamicValue|null;
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts]: Per below, let's keep this as `Expression`.

File: packages/compiler/src/jit_compiler_facade.ts [hunk 3]
```diff
@@ -91,6 +91,7 @@ export class CompilerFacadeImpl implements CompilerFacade {
       emitInline: true,
       containsForwardDecls: false,
       schemas: facade.schemas ? facade.schemas.map(wrapReference) : null,
+      id: facade.id || null,
```
Comment (alxhub) [inline packages/compiler/src/jit_compiler_facade.ts]: We can wrap `facade.id` in a `WrappedNodeExpr` here.

**Review by alxhub — CHANGES_REQUESTED** [review]

**Comment by crisbeto** [comment]
@alxhub It's been reworked based on the feedback.

File: packages/compiler/src/compiler_facade_interface.ts [hunk 4]
```diff
@@ -106,6 +106,7 @@ export interface R3NgModuleMetadataFacade {
   exports: Function[];
   emitInline: boolean;
   schemas: {name: string}[]|null;
+  id: any;
```
Comment (alxhub) [inline packages/compiler/src/compiler_facade_interface.ts]: It can remain a `string` here :)

**Review by alxhub — APPROVED** [review]
One comment, otherwise LGTM!

**Comment by kara** [comment]
[presubmit](https://tap.corp.google.com/ui#id=OCL:246937836:BASE:246937916:1557192817977:30c936af)

**Comment by kara** [comment]
This is failing Google presubmit because of a cyclical dependency:

```
angular2/rc/packages/core/src/linker/ng_module_factory_loader.ts -> 
angular2/rc/packages/core/public_api.ts -> angular2/rc/packages/core/src/core.ts -> 
angular2/rc/packages/core/src/metadata.ts -> angular2/rc/packages/core/src/metadata/directives.ts -> 
angular2/rc/packages/core/src/render3/jit/directive.ts -> 
angular2/rc/packages/core/src/render3/jit/environment.ts -> 
angular2/rc/packages/core/src/render3/index.ts -> 
angular2/rc/packages/core/src/render3/ng_module_ref.ts -> 
angular2/rc/packages/core/src/linker/ng_module_factory_loader.ts
```

**Comment by alxhub** [comment]
We can remove `registerNgModuleType` from `jit/environment.ts` as it's no longer generated by the compiler (also remove it from `r3_identifiers.ts`). That will break the circular dependency.

**Comment by crisbeto** [comment]
Thank you for the pointer @alxhub! This is ready for another presubmit.

**Comment by kara** [comment]
[presubmit](http://test/OCL:246937836:BASE:247080468:1557261253192:9d545fd0)

**Comment by kara** [comment]
Still getting TAP failures:

```
/angular2/rc/packages/core/src/linker/ng_module_factory_loader.ts -> 
/angular2/rc/packages/core/public_api.ts -> 
/angular2/rc/packages/core/src/core.ts -> 
/angular2/rc/packages/core/src/metadata.ts -> 
/angular2/rc/packages/core/src/metadata/directives.ts -> 
/angular2/rc/packages/core/src/render3/jit/directive.ts -> 
/angular2/rc/packages/core/src/render3/jit/environment.ts -> 
/angular2/rc/packages/core/src/render3/index.ts -> 
/angular2/rc/packages/core/src/render3/ng_module_ref.ts -> 
/angular2/rc/packages/core/src/linker/ng_module_factory_loader.ts
```

**Comment by crisbeto** [comment]
Moved some things around based on our discussion with @alxhub. Can you take another look? It's all in the second `fixup` commit.

**Comment by kara** [comment]
[new presubmit](http://test/OCL:247929583:BASE:247929597:1557757256999:b9e10469)

**Comment by kara** [comment]
merge assistance: change made to Ivy-specific API

**Pull request closed without merging**